packages/git/src/browser/git-decoration-provider.ts

```diff
@@ -14,54 +14,103 @@
 // SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
 // *****************************************************************************
 
-import { inject, injectable } from '@theia/core/shared/inversify';
+import { inject, injectable, postConstruct } from '@theia/core/shared/inversify';
 import { GitFileChange, GitFileStatus, GitStatusChangeEvent } from '../common';
 import { CancellationToken, Emitter, Event } from '@theia/core/lib/common';
 import { Decoration, DecorationsProvider } from '@theia/core/lib/browser/decorations-service';
 import { GitRepositoryTracker } from './git-repository-tracker';
 import URI from '@theia/core/lib/common/uri';
+import { GitConfiguration, GitPreferences } from './git-preferences';
+import { PreferenceChangeEvent } from '@theia/core/lib/browser';
 
 @injectable()
 export class GitDecorationProvider implements DecorationsProvider {
 
+    @inject(GitPreferences) protected readonly preferences: GitPreferences;
+    @inject(GitRepositoryTracker) protected readonly gitRepositoryTracker: GitRepositoryTracker;
+
+    protected decorationsEnabled: boolean;
+    protected colorsEnabled: boolean;
+
+    protected decorations = new Map<string, Decoration>();
+    protected uris: Set<string> = new Set<string>();
+
+    /**
+     * Cached change event for re-rendering decorations.
+     */
+    protected changeEvent: GitStatusChangeEvent | undefined;
```

Comment:
Suggestion: I'm not sure whether we need to store the `changeEvent` if we already store the `uris` and `decorations`. We can simply recompute `uris` and `decorations` when a `GitStatusChangeEvent` occurs and reuse these results.

Reply:
I've got to admit, I didn't think about that. I've provided a patch that shows a working solution: [decorations-patch.txt](https://github.com/eclipse-theia/theia/files/9585579/decorations.txt)

I'm not sure whether that's better though, since it introduces some complexity into `provideDecorations` (but removes some complexity from `collectDecorationData`).